### PR TITLE
feat: Support Gradle projects

### LINKED ISSUE #67: Add gradle support
Once @aloubyansky and team are done with the next iteration of Gradle improvements, in particular the ability to add and list extensions.
We should add gradle support for the launcher.

At this stage, we will need to decide:

* how "gradle create" will be done (via a command too or from a separate lib generating this shell)
* whether the `mvnw` and `gradlew` should be created by a shared library instead of int he launcher itself.

### PR DESCRIPTION
**Pull request by gastaldi**
This adds a combobox in the UI to select the preferred Build tool

Fixes #67 
 
Screenshot from 2019-10-03 17-39-26

### CODE REVIEW

**Comment by aloubyansky** [comment]
Is this some kind of formatting related failure?

**Comment by maxandersen** [comment]
Awesome! A few observations:

1) The combo box looks kinda out of place - any way to make it have more the same style ? Or maybe should just for now have it be a "Use Gradle: [ ]" option with a the green styled label ?

2) When you download the zip a dialog pops up explaining you to use maven - I don't see changes in here related to that dialog box to also mention gradle ?

3) I'm currently working on fixing an issue of gradlew not being included in the generated result - it is not happening if gradle is missing or if java_home not set in global context. Does the generated result actually have the gradle wrapper included when deployed as app ?

**Comment by gastaldi** [comment]
Screenshot updated

**Comment by maxandersen** [comment]
great - then my #1 is fixed and can be ignored ;)

**Comment by gastaldi** [comment]
> When you download the zip a dialog pops up explaining you to use maven - I don't see changes in here related to that dialog box to also mention gradle ?

Good catch! Fixing it

> I'm currently working on fixing an issue of gradlew not being included in the generated result - it is not happening if gradle is missing or if java_home not set in global context. Does the generated result actually have the gradle wrapper included when deployed as app ?

@maxandersen not yet. I was expecting your fix :smile:

**Comment by gastaldi** [comment]
@maxandersen fixed your item 2 with https://github.com/quarkusio/code.quarkus.io/pull/163/commits/64d26a36c8d740f3dd01dae5d4c57b722644f2bb

**Comment by maxandersen** [comment]
> @maxandersen not yet. I was expecting your fix 😄

don't get your hopes up :) gradle is exploding on me when I try to give it JAVA_HOME :)

note, afaics it will actually work as long as users shell/profile has JAVA_HOME + gradle installed. Its only when you use something like SDKman that sets things up locally rather than globally things gets weird.

File: src/main/kotlin/io/quarkus/code/QuarkusProjectCreator.kt [hunk 1]
```diff
@@ -30,40 +38,52 @@ open class QuarkusProjectCreator {
...
-        writeResourceFile(zipWrite, MVNW, true)
+        zipWrite.mkdirs(MVNW_WRAPPER_DIR)
+        writeResourceFile(zipWrite, MVNW_RESOURCES_DIR, MVNW_WRAPPER_JAR)
+        writeResourceFile(zipWrite, MVNW_RESOURCES_DIR, MVNW_WRAPPER_PROPS)
+        writeResourceFile(zipWrite, MVNW_RESOURCES_DIR, MVNW_WRAPPER_DOWNLOADER)
+        writeResourceFile(zipWrite, MVNW_RESOURCES_DIR, MVNW_CMD, true)
+        writeResourceFile(zipWrite, MVNW_RESOURCES_DIR, MVNW, true)
+    }
+
+    private fun addGradlew(zipWrite: CommonsZipProjectWriter) {
```
Comment (maxandersen) [inline src/main/kotlin/io/quarkus/code/QuarkusProjectCreator.kt]: wait - code.quarkus.io does its own writing and creation of the mvn/gradle wrappers ? 
we should put this stuff into the base tooling rather than code.quarkus.io duplicates it.

for this specific issue of getting gradlew enabled on code.quarkus.io + 1 though. just something we should get cleaned up.

Comment (gastaldi) [reply src/main/kotlin/io/quarkus/code/QuarkusProjectCreator.kt]: I remember @ia3andy pointing an issue where the Maven wrapper wasn't created in the devtools-common, but I can't find the issue, maybe that's already fixed

Comment (maxandersen) [reply src/main/kotlin/io/quarkus/code/QuarkusProjectCreator.kt]: Yes, and it is still not created there - and the mvn mojo does call outs to mvn/gradle with "random" success - I found that the other day and started to wonder what kind of funky differences we have :)

But yeah, I think we should move this code into devtools-common and tell users to run `gradle wrapper --upgrade` and if similar exist for `mvn` wrapper.

Comment (gastaldi) [reply src/main/kotlin/io/quarkus/code/QuarkusProjectCreator.kt]: Agreed, feel free to copy those files over to devtools-common, since you're already working on it :wink:

Comment (gastaldi) [reply src/main/kotlin/io/quarkus/code/QuarkusProjectCreator.kt]: Steps to upgrade the `mvn`wrapper: https://github.com/takari/takari-maven-plugin#updating-the-maven-wrapper-in-your-maven-project

TLDR: `mvn -N io.takari:maven:0.7.6:wrapper` and it will overwrite the installed files

Comment (gastaldi) [reply src/main/kotlin/io/quarkus/code/QuarkusProjectCreator.kt]: Found the issue: https://github.com/quarkusio/code.quarkus.io/issues/15

**Comment by ia3andy** [comment]
This looks great @gastaldi :) 

I'll test it locally, add a few frontend tests and merge.

PS: I really need to introduce the mock deployment in PRs and test coverage like we had in the launcher frontend

**Comment by ia3andy** [comment]
@gastaldi never mind the frontend tests you already did.. you always impress me 😁

**Comment by ia3andy** [comment]
@gastaldi in the UI, I see a few problems:
1. the "Generate your application" button is not aligned correctly (it should be at the bottom)
2. the black arrow of the select box should be red or blue (#ff004a or #4695eb) I am not sure which would fit best
3. it looks a bit weird to have the whole panel take all that space when the option box is closed (but this will be automatically fixed in a soon coming PR that will show the Quarkus version as a readonly select box)

**Comment by ia3andy** [comment]
The padding in the bottom is a bit too large:
image

**Comment by gastaldi** [comment]
@ia3andy I just pushed a commit that may have solved the button alignment issue

image

**Comment by ia3andy** [comment]
The copy to clipboard extension cmd is now contextual to the build tool :)
image

**Comment by gastaldi** [comment]
@ia3andy good catch! Forgot about that

**Comment by ia3andy** [comment]
- I've moved the Build Tool to the base options and added the Quarkus version in the extended options.
- I've added a readonly flag on the Quarkus Version

image

File: src/main/kotlin/io/quarkus/code/CodeQuarkusResource.kt [hunk 2]
```diff
@@ -139,11 +139,12 @@ class CodeQuarkusResource {
             extensions: Set<String>,
 
             @DefaultValue("MAVEN")
+            @NotEmpty
             @QueryParam("b")
-            @Parameter(name = "b", description = "The build tool to use (MAVEN or GRADLE) (default: MAVEN)")
-            buildTool: io.quarkus.code.model.BuildTool
```
Comment (gastaldi) [inline src/main/kotlin/io/quarkus/code/CodeQuarkusResource.kt]: You can remove the `io.quarkus.code.model.BuildTool` class now

**Comment by ia3andy** [comment]
The Quarkus Version has a contextual tooltip:
You may change the Quarkus Version after generation in the ${configFileName}. Just be cautious with extension compatibily.

**Comment by gastaldi** [comment]
Perhaps it would be better to replace the `READ-ONLY` indicator with an icon,  something like: 
image

File: src/main/kotlin/io/quarkus/code/CodeQuarkusResource.kt (hunk 2 above)
Comment (ia3andy) [reply src/main/kotlin/io/quarkus/code/CodeQuarkusResource.kt]: Good catch. I hesitated for a while before not using it, but I found it more confusing since there is one declared in Quarkus already.

The second reason is stupid, I will not say it 😅

**Comment by ia3andy** [comment]
@gastaldi do you mind giving it a try before we merge?

Are you happy with the end result (beside your note on the readonly)?

**Comment by ia3andy** [comment]
@gastaldi I noticed the generated project contains twice the Quarkus version, once in the gradle.properties and then in the settings.gradle `useModule("io.quarkus:quarkus-gradle-plugin:0.24.0")` is it expected?

**Comment by gastaldi** [comment]
@ia3andy I have no idea, that's the output of the CreateProject class. Perhaps @geoand may know more about it?

**Comment by ia3andy** [comment]
about the readonly flag, I am going to ask JasonB

**Comment by gastaldi** [comment]
@jkbrock :point_up:

**Comment by ia3andy** [comment]
ok @gastaldi design has been approved, here is what we landed on:
image

**Comment by ia3andy** [comment]
We found a BUG in the generated project. I will investigate more tomorrow, until then this PR is blocked!

**Comment by ia3andy** [comment]
When editing project info, the resulting project contains empty files:
- gradle.properties
- settings.gradle

**Comment by ia3andy** [comment]
ON HOLD until https://github.com/quarkusio/quarkus/issues/4601 is fixed and released.
